### DISCUSSION TITLE
How to get instance of a class inside a hooked method?

### DISCUSSION THREAD

**Discussion by afk-Legacy** [vfsfitvnm/frida-il2cpp-bridge#114] [Q&A]
Hey, thanks for this awesome project! I had a question regarding getting the instance of a class of which I have hooked a function, the main reason of getting instance is so that I can modify the variables present in the class. 

Heres what I have tried so far:
(Do you have any social media / discord for queries/quesiton?)
```

import "frida-il2cpp-bridge";

export function log(message: String) : void 
{
    (globalThis as any).console.log(message);
}


Il2Cpp.perform(() => {
    // it will use default directory path and file name: /<default_path>/<default_name>.cs
    //Il2Cpp.Dumper.classicDump("Dump.cs", "/sdcard/Download/");


    const AssemblyCSharp = Il2Cpp.Domain.assemblies["Assembly-CSharp"].image;
    const SaveManager = AssemblyCSharp.classes["SaveManager"];

    const UpdateState = SaveManager.methods["Load"];
    var shouldLog = true;

    UpdateState.implementation = function () : void
        {
            const instance = this.handle.readPointer();
            
            const SaveState = instance.add("0xC");
            
            log("Game Version: " + SaveState.add(0x8).readFloat());

            this.methods.UpdateState.invoke<void>();
        };


});
```

**Comment by vfsfitvnm** [comment]
Please read [here](https://github.com/vfsfitvnm/frida-il2cpp-bridge#methods).
However it doesn't make sense "instance of a class" if the method you are hooking is a static method.
If you want to edit/call static fields/methods, you don't need to hook.
```ts
const SaveManager = AssemblyCSharp.classes["SaveManager"];

SaveManager.fields.whatever.value = ...;
SaveManager.methods.set_Whatever.invoke(...);
```

> Do you have any social media / discord for queries/quesiton?

Just open a new issue here

**Comment by afk-Legacy** [comment]
@vfsfitvnm Its not a static method.

**Comment by vfsfitvnm** [comment]
> @vfsfitvnm Its not a static method.

Then `this` will be a `Il2Cpp.Object`

**Comment by afk-Legacy** [comment]
Alright, so to get the address of the field, I could do something like:

```
UpdateState.implementation = function () : void
        {
            const instance = this.handle.readPointer();
            
            const SaveState = this.feilds.state;

            log("SaveState Address: " + SaveState.handle.readPointer());

            this.methods.UpdateState.invoke<void>();
        };
```
        
        SaveState is reference to another class whose field I wish to access, so after getting the address of SaveState I could add the required offset to access another field present in the class of SaveState.

**Comment by vfsfitvnm** [comment]
But why are you going low level?
Accessing to fields is already well covered.
```ts
UpdateState.implementation = function () : void {
    ((this as Il2Cpp.Object).fields.whatever.value as Il2Cpp.Object).fields.whatever2.value = ...;
    this.methods.UpdateState.invoke<void>();
};
```

**Comment by afk-Legacy** [comment]
The field I want to access is this: 
![image](https://user-images.githubusercontent.com/83068093/133434506-6310b30d-0a04-4784-b764-f47866dad887.png)

I want to access it from another class, i.e from the class where it is being loaded/saved from, to do so I find the reference to SaveState Class in SaveManager class.

![image](https://user-images.githubusercontent.com/83068093/133434702-71400464-dad3-4eda-bbcd-b8f840316aa8.png)


I am new to this so excuse my inability/ignorance. Normally in hooking I would get the instance of SaveManager class, add 0xC in it to get the reference class (SaveState) and add the desired offset of playername field (0xC) to get the player name. I intend to do the same here.

**Comment by afk-Legacy** [comment]
Ah I did not see your message, it works perfectly! Thanks...

**Comment by vfsfitvnm** [comment]
Oh ok no problem.
> Normally in hooking I would get the instance of SaveManager class, add 0xC in it to get the reference class (SaveState) and add the desired offset of playername field (0xC) to get the player name.

This is exactly what it does under the hood. You can still use offsets and pointers, of course.

```ts
const SaveManager = AssemblyCSharp.classes.SaveManager;

SaveManager.methods.UpdateState.implementation = function (): void {
    const state = (this as Il2Cpp.Object).fields.state.value as Il2Cpp.Object;
    const playerName = state.fields.playerName.value as Il2Cpp.String;
    
    log(`Player name: ${playerName}`);
    
    this.methods.UpdateState.invoke<void>();
};
```
is equivalent to:
```ts
const SaveManager = AssemblyCSharp.classes.SaveManager;

SaveManager.methods.UpdateState.implementation = function (): void {
    const state = this.handle.add(0xc).readPointer();
    const playerName = state.add(0xc).readPointer();
    
    log(`Player name: ${playerName.add(Process.pointerSize * 2 + 4).readUtf16String()}`);
    
    this.methods.UpdateState.invoke<void>();
};
```

**Comment by vfsfitvnm** [comment]
PS: the "hardcoded" solution is way faster, of course, but it's not portable.
If you need the best performance without compromising portability (and high level abstraction), you can do the following:

```ts
const SaveManager = AssemblyCSharp.classes.SaveManager;
const SaveState = AssemblyCSharp.classes.SaveState;

SaveManager.methods.UpdateState.implementation = function (): void {
    const state = this.handle.add(SaveManager.fields.state.offset).readPointer();
    const playerNameRaw = state.add(SaveState.fields.playerName.offset).readPointer();
    const playerName = new Il2Cpp.String(playerName);

    log(`Player name: ${playerName}`);
    
    this.methods.UpdateState.invoke<void>();
};
```
